examples/cmd/pool.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/src-d/metadata-retrieval/github"
+)
+
+type DownloadersPool struct {
+	Size    int
+	pool    chan *DownloaderPoolItem
+	started bool
+	ended   bool
+	t0      time.Time
+	t1      time.Time
+	stats0  map[int]*downloaderStats
+	stats1  map[int]*downloaderStats
+}
+
+type DownloaderPoolItem struct {
+	id int
+	*github.Downloader
+}
+
+type DownloaderPoolStats struct {
+	Elapsed    time.Duration
+	RatesUsage map[int]*RateUsage
+}
+
+func (d *DownloaderPoolItem) stats() (*downloaderStats, error) {
+	rate, err := d.RateRemaining(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	return &downloaderStats{
+		Rate: rate,
+		Time: time.Now(),
+	}, nil
+}
+
+type downloaderStats struct {
+	Rate int
+	Time time.Time
+}
+
+type downloaderBuilder = func(c *http.Client) (*github.Downloader, error)
+
+func NewDownloadersPool(clients []*http.Client, builder downloaderBuilder) (*DownloadersPool, error) {
+	ch := make(chan *DownloaderPoolItem, len(clients))
+
+	for i, c := range clients {
+		d, err := builder(c)
```

Comment:
if this construction function isn't responsible for creating downloaders what is the advantage of accepting clients+builder instead of just an array of downloaders?

Reply:
No advantage actually. Most probably this was different in some initial version I wrote and then I just  didn't rethink it, thanks!